pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -301,15 +320,87 @@ func (e *TokensController) syncSecret() {
 				retry = retriable
 			}
 		default:
-			// Update token if needed
-			if retriable, err := e.generateTokenIfNeeded(logger, sa, secret); err != nil {
-				logger.Error(err, "Populating serviceaccount token", "secret", klog.KRef(secretInfo.namespace, secretInfo.name), "serviceAccount", klog.KRef(secretInfo.namespace, secretInfo.saName))
-				retry = retriable
+			// Check if the legacy service account token can be cleaned up
+			if e.shouldCleanToken(secret, sa) {
+				// Delete token
+				klog.V(4).Infof("syncSecret(%s/%s), auto-generated token has not been used for more than 366 * 24 hours, deleting token", secretInfo.namespace, secretInfo.name)
+				if retriable, err := e.deleteToken(secretInfo.namespace, secretInfo.name, secretInfo.uid); err != nil {
+					klog.Errorf("error deleting auto-generated serviceaccount token %s/%s for service account %s: %v", secretInfo.namespace, secretInfo.name, secretInfo.saName, err)
+					retry = retriable
+				}
+			} else {
+				// Update token if needed
+				if retriable, err := e.generateTokenIfNeeded(logger, sa, secret); err != nil {
+					logger.Error(err, "Populating serviceaccount token", "secret", klog.KRef(secretInfo.namespace, secretInfo.name), "serviceAccount", klog.KRef(secretInfo.namespace, secretInfo.saName))
+					retry = retriable
+				}
 			}
 		}
 	}
 }
 
+func (e *TokensController) shouldCleanToken(secret *v1.Secret, sa *v1.ServiceAccount) bool {
+	if !e.legacyTokenCleanUp {
+		return false
+	}
+
+	// determine whether the secret is referenced by the service account, if not referenced, the secret is manually generated and should not be cleaned up
+	if !getSecretReferences(sa).Has(secret.Name) {
+		return false
+	}
+
+	// determine the trackedSince time
+	configMap, err := e.client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(context.TODO(), legacytokentracking.ConfigMapName, metav1.GetOptions{})
```

Comment:
I would have guessed we'd have an informer-driven way to get this configmap so you don't need to do a live lookup on every sync of every token secret

watching that configmap should let us keep updated a member field that indicates whether we're even in cleanup mode or not, and if so, what the maximum last-used date we're cleaning up is

we should be able to check that member field here without doing a live configmap lookup

Reply:
Hi Jordan, may I know why we need to keep updating a member field? Can I just maintain a filtered configmap informer which only caches the configmap we want and read the tracked_since value from the cached configmap every time we need?